paimon-common/src/main/java/org/apache/paimon/lookup/sort/SortLookupStoreWriter.java

```diff
@@ -0,0 +1,212 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.lookup.sort;
+
+import org.apache.paimon.compression.BlockCompressionFactory;
+import org.apache.paimon.compression.BlockCompressionType;
+import org.apache.paimon.compression.BlockCompressor;
+import org.apache.paimon.lookup.LookupStoreFactory;
+import org.apache.paimon.lookup.LookupStoreWriter;
+import org.apache.paimon.memory.MemorySegment;
+import org.apache.paimon.memory.MemorySlice;
+import org.apache.paimon.options.MemorySize;
+import org.apache.paimon.utils.BloomFilter;
+import org.apache.paimon.utils.MurmurHashUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.zip.CRC32;
+
+import static org.apache.paimon.lookup.sort.BlockHandle.writeBlockHandle;
+import static org.apache.paimon.memory.MemorySegmentUtils.allocateReuseBytes;
+import static org.apache.paimon.utils.VarLengthIntUtils.encodeInt;
+
+/** A {@link LookupStoreWriter} for sorting. */
+public class SortLookupStoreWriter implements LookupStoreWriter {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(SortLookupStoreWriter.class.getName());
+
+    public static final int MAGIC_NUMBER = 1481571681;
+
+    private final BufferedOutputStream fileOutputStream;
+    private final int blockSize;
+    private final BlockWriter dataBlockWriter;
+    private final BlockWriter indexBlockWriter;
+    @Nullable private final BloomFilter.Builder bloomFilter;
+    private final BlockCompressionType compressionType;
+    @Nullable private final BlockCompressor blockCompressor;
+
+    private byte[] lastKey;
+    private long position;
+
+    private long recordCount;
+    private long totalUncompressedSize;
+    private long totalCompressedSize;
+
+    SortLookupStoreWriter(
+            File file,
+            int blockSize,
+            @Nullable BloomFilter.Builder bloomFilter,
+            @Nullable BlockCompressionFactory compressionFactory)
+            throws IOException {
+        this.fileOutputStream = new BufferedOutputStream(Files.newOutputStream(file.toPath()));
+        this.blockSize = blockSize;
+        this.dataBlockWriter = new BlockWriter((int) (blockSize * 1.1));
+        int expectedNumberOfBlocks = 1024;
```

Comment:
Currently, the default file size for primary key tables is 128MB. If we use 64KB as the block size, the number of blocks might be around 2048. So, could the expected number of blocks be 2048?

Reply:
Perhaps even larger, currently it is a column storage file converted to row storage